Check RegisterMetricAndTrackRateLimiterUsage error when starting BootstrapSigner & TokenCleaner controllers

frodenas opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
Prevent `BootstrapSigner` and `TokenCleaner` controllers to start if `metrics.RegisterMetricAndTrackRateLimiterUsage` returns an error.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: complements #53571 

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

xiangpengzhao commented:
/ok-to-test

frodenas commented:
/test pull-kubernetes-kubemark-e2e-gce

frodenas commented:
/assign @deads2k @mikedanese

deads2k commented:
It doesn't seem like a failure to add metrics should result not running the controller. I could go as far as a log message if you like.

frodenas commented:
@deads2k 

1. Before #53571 was merged, an error registering a metric panicked the controller at startup time (see example at #52801), so this PR doesn't change the behavior in the sense that the controller is not started if there's a problem.
2. A log message can go unnoticed. When there's a problem and you need to look at your prometheus metrics, then you realize that you don't have metrics. IMO, I don't believe this is acceptable.
3. Errors registering a metric at the [metrics.RegisterMetricAndTrackRateLimiterUsage](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/metrics/util.go#L40) call usually happen because a contributor (not an end-user): a) created/updated a controller and specified an invalid `owner name`(ie not compliant with [prometheus metric name rules](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels) and what happened at [#52801](https://github.com/kubernetes/kubernetes/issues/52801#issuecomment-331286326)) b) created/updated a controller and used an `owner name` already registered. Better to catch up those errors during development/test time that not users experiencing this problem.

Thoughts?

deads2k commented:
> Errors registering a metric at the metrics.RegisterMetricAndTrackRateLimiterUsage call usually happen because a contributor (not an end-user): a) created/updated a controller and specified an invalid owner name(ie not compliant with prometheus metric name rules and what happened at #52801) b) created/updated a controller and used an owner name already registered. Better to catch up those errors during development/test time that not users experiencing this proble

Ok.  That makes sense.

Please combine with https://github.com/kubernetes/kubernetes/pull/53591

deads2k commented on pkg/controller/bootstrap/bootstrapsigner_test.go:
```diff
@@ -60,13 +64,13 @@ func newConfigMap(tokenID, signature string) *v1.ConfigMap {
 
 func TestNoConfigMap(t *testing.T) {
-	signer, cl := newBootstrapSigner()
+	signer, cl, _ := newBootstrapSigner()
```
don't ignore the error.

frodenas replied:
ok

frodenas commented:
/test pull-kubernetes-unit

deads2k commented on cmd/kube-controller-manager/app/bootstrap.go:
```diff
@@ -16,20 +16,32 @@ limitations under the License.
-	).Run(ctx.Stop)
+	)
+	if err != nil {
+		return false, fmt.Errorf("error creating bootstrap signer controller: %v", err)
```
`return true, ....`

deads2k commented on cmd/kube-controller-manager/app/bootstrap.go:
```diff
@@ -16,20 +16,32 @@ limitations under the License.
-	).Run(ctx.Stop)
+	)
+	if err != nil {
+		return false, fmt.Errorf("error creating token cleaner controller: %v", err)
```
`return true, ....`

deads2k commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -256,7 +256,9 @@ func startResourceQuotaController(ctx ControllerContext) (bool, error) {
 	if resourceQuotaControllerClient.CoreV1().RESTClient().GetRateLimiter() != nil {
-		metrics.RegisterMetricAndTrackRateLimiterUsage("resource_quota_controller", resourceQuotaControllerClient.CoreV1().RESTClient().GetRateLimiter())
+		if err := metrics.RegisterMetricAndTrackRateLimiterUsage("resource_quota_controller", resourceQuotaControllerClient.CoreV1().RESTClient().GetRateLimiter()); err != nil {
+			return false, err
```
`return true, ....`

frodenas replied:
Why `true` when some lines below is returning `false`? https://github.com/frodenas/kubernetes/blob/04d400c29ba2b5588ceb548473e60e0ad2ea7721/cmd/kube-controller-manager/app/core.go#L266

Also, this `bool` indicates if the controller has been [started](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L478), and it's failing before the `Run`, so it should return `false` right?

deads2k replied:
> Also, this bool indicates if the controller has been started, and it's failing before the Run, so it should return false right?

The bool indicates that it was enabled.  Some controllers have secondary enablement criteria (like API availability) that they own inside their own start methods.

deads2k commented on pkg/controller/bootstrap/bootstrapsigner_test.go:
```diff
@@ -60,13 +64,19 @@ func newConfigMap(tokenID, signature string) *v1.ConfigMap {
-	signer, cl := newBootstrapSigner()
+	signer, cl, err := newBootstrapSigner()
+	if err != nil {
+		t.Errorf("Error creating BootstrapSigner: %v", err)
```
`t.Fatal`.  the rest of the test won't work.

deads2k commented:
minor comments.  squash, then lgtm

frodenas commented:
@deads2k ok, will update it. Question: #53593, #53594, #53595. #53643, #53646 and #53648 are related. Do you want them combined in this PR (similar to #53591)? If yes, I'll pick up and fix those commits with your comments in this PR.

deads2k commented:
> @deads2k ok, will update it. Question: #53593, #53594, #53595. #53643, #53646 and #53648 are related. Do you want them combined in this PR (similar to #53591)? If yes, I'll pick up and fix those commits with your comments in this PR.

Yeah, gather them all here.

frodenas committed:
Add metrics.UnregisterMetricAndUntrackRateLimiterUsage function

For testing purposes, we want to unregister a previously registered
rate limiter prometheus metric and stop the goroutine that updates
this metric.

Signed-off-by: Ferran Rodenas <rodenasf@vmware.com>

frodenas committed:
Check RegisterMetricAndTrackRateLimiterUsage error when starting controllers

Signed-off-by: Ferran Rodenas <rodenasf@vmware.com>

deads2k commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -68,8 +68,7 @@ func startServiceController(ctx ControllerContext) (bool, error) {
 	if err != nil {
-		glog.Errorf("Failed to start service controller: %v", err)
-		return false, nil
+		return true, fmt.Errorf("error creating Service controller: %v", err)
```
This one shouldn't fail.  I don't recall the details, but this was made a glog for some reason.  It lives like this as a legacy.

frodenas replied:
let me revert back this change and I will add a comment

deads2k commented:
/lgtm

but I'm guessing something is double registering....

frodenas commented:
/test pull-kubernetes-unit

frodenas commented:
@deads2k PTAL, had to rebase and updated a new [test](https://github.com/kubernetes/kubernetes/tree/master/test/integration/daemonset) added upstream. Thanks.

deads2k commented:
/lgtm

frodenas commented:
/assign @brendandburns @lavalamp @smarterclayton

brendandburns commented:
/unassign

frodenas commented:
/test pull-kubernetes-unit

deads2k commented:
hanging out because of pkg/util/metrics.  That lives in service this.  tagging.
